src/image/VisitInfo.cc

```diff
@@ -247,71 +255,73 @@ class VisitInfoSchema {
-                      "refracted apparent topocentric position of boresight at middle of exposure", "")),
-              boresightAirmass(schema.addField<double>(
-                      "boresightairmass", "airmass at boresight, relative to zenith at sea level", "")),
-              boresightRotAngle(schema.addField<lsst::geom::Angle>(
-                      "boresightrotangle", "rotation angle at boresight at middle of exposure", "")),
-              rotType(schema.addField<int>("rottype", "rotation type; see VisitInfo.getRotType for details",
-                                           "MJD")),
-
-              // observatory data
-              latitude(schema.addField<lsst::geom::Angle>(
-                      "latitude", "latitude of telescope (+ is east of Greenwich)", "")),
-              longitude(schema.addField<lsst::geom::Angle>("longitude", "longitude of telescope", "")),
-              elevation(schema.addField<double>("elevation", "elevation of telescope", "")),
-
-              // weather data
-              airTemperature(schema.addField<double>("airtemperature", "air temperature", "C")),
-              airPressure(schema.addField<double>("airpressure", "air pressure", "Pascal")),
-              humidity(schema.addField<double>("humidity", "humidity (%)", "")),
-
-              instrumentLabel(schema.addField<std::string>(
-                      "instrumentlabel", "Short name of the instrument that took this data", "", 0)),
-
-              // for internal support
-              version(schema.addField<int>(VERSION_KEY, "version of this VisitInfo")),
-
-              idnum(schema.addField<table::RecordId>("idnum", "identifier of this full focal plane exposure",
-                                                     "")),
-
-              focusZ(schema.addField<double>("focusz", "defocal distance", "mm")),
-
-              observationType(schema.addField<std::string>(
-                      "observationType", "type of this observation (e.g. science, flat, bias)", "", 0)),
-              scienceProgram(schema.addField<std::string>(
-                      "scienceProgram", "observing program (survey or proposal) identifier", "", 0)),
-              observationReason(schema.addField<std::string>(
-                      "observationReason", "reason this observation was taken, or its purpose", "", 0)),
-              object(schema.addField<std::string>("object", "object of interest or field name", "", 0)),
-              hasSimulatedContent(schema.addField<afw::table::Flag>(
-                      "hasSimulatedContent", "Was any part of this observation simulated?")) {}
+    VisitInfoSchema(int _version) : schema() {
+        if (_version < 5) {
+            // exposureId was removed in version 5, but we need to reserve space in the schema for old files.
+            // We don't save the field record, as we don't need to read into it.
+            schema.addField<table::RecordId>("exposureid", "exposure ID", "");
+        }
```

Comment:
This seems... not very portable if there are more schema changes in the future; the logic for exactly which fields belong in which version could get quite twisty. You also have to sacrifice the benefits of one-time initialization to do it this way. And then there's the potential confusion from `VisitInfoSchema` acting like a singleton but not actually being one (a dualton?).

Why not just have separate classes for versions 0-4 and 5+, and make `VisitInfoFactory` responsible for initializing the correct schema? Then you could extend to future incompatibilities without worrying about accidentally changing the old stuff.

Reply:
I'd tried your latter approach, but ran into C++ compiler problems that I couldn't sort out. This approach was @TallJimbo 's suggest (on top of my initial attempt that forgot about the static members and had the "get which version" in the factory).

Line number no longer accurate, but this was the error (I had made a `VisitInfoSchemaWithExposureId` subclass):
```
src/image/VisitInfo.cc:322:39: error: member initializer 'schema' does not name a non-static data member or base class
    VisitInfoSchemaWithExposureId() : schema() {
```

Reply:
🤦 Right, I've been working in Python for too long. But this version's also potentially unsafe re: its API (you're fine this time because you don't care about the expId key in either case), so maybe file a ticket to clean up the implementation later?

Reply:
I'm game to try another approach here, if you have a suggestion (or a way around the error above, which I think is due to static-ness of `schema` shared between the parent and child classes). I think CRTP would be one way around it.

Reply:
I don't understand why `schema` would be `static` (it's not so in the current or previous versions?), but I don't think this is a problem we can solve in <1 hour. The current code at least gives the desired behavior for now.

Reply:
https://jira.lsstcorp.org/browse/DM-39906